apis/python/src/tiledbsoma/experiment_query/query.py

```diff
@@ -0,0 +1,503 @@
+                    column_names=["soma_joinid"],
+                )
+                .concat()
+                .column("soma_joinid")
+                .combine_chunks()
+            )
+        return self._joinids[axis]
+
+    def _load_joinids(self) -> None:
+        """Private. Ensure joinids for both axis are in-memory."""
+        futures = []
+        if not self._joinids["obs"]:
+            futures.append(self._threadpool.submit(self.obs_joinids))
+        if not self._joinids["var"]:
+            futures.append(self._threadpool.submit(self.var_joinids))
+        if futures:
+            concurrent.futures.wait(futures)
+        assert self._joinids["obs"] is not None, "Internal error"
+        assert self._joinids["var"] is not None, "Internal error"
+
+    def _axisp_inner(
+        self,
+        axis: Literal["obs", "var"],
+        layer: str,
+    ) -> SparseNDArrayRead:
+        assert axis in ["obs", "var"]
+        key = f"{axis}p"
+
+        ms = self.experiment.ms[self.measurement_name]
+        if key not in self.experiment.ms[self.measurement_name]:
+            raise ValueError(f"Measurement does not contain {key} data")
+
+        axisp = ms.obsp if axis == "obs" else ms.varp
+        if not (layer and layer in axisp):
+            raise ValueError(f"Must specify '{key}' layer")
+
+        if axisp[layer].soma_type != "SOMASparseNDArray":
+            raise TypeError(f"Unexpected SOMA type stored in '{key}' layer")
+        assert isinstance(axisp[layer], SOMASparseNDArray)
+
+        self._load_joinids()
+        assert self._joinids[axis] is not None
+
+        joinids = self._joinids[axis]
+        return axisp[layer].read((joinids, joinids))
+
+    def _rewrite_X_for_positional_indexing(
+        self, X_tables: Dict[str, pa.Table]
+    ) -> Dict[str, pa.Table]:
+        """
+        Private convenience function to convert axis dataframe to X matrix joins
+        from `soma_joinid`-based joins to positionally indexed joins (like AnnData uses).
+
+        Input is organized as:
+            obs[i] annotates X[ obs[i].soma_joinid, : ]
+        and
+            var[j] annotates X[ :, var[j].soma_joinid ]
+
+        Output is organized as:
+            obs[i] annotates X[i, :]
+        and
+            var[j] annotates X[:, j]
+
+        In addition, the `soma_joinid` column is dropped from the axis dataframes.
+        """
+        new_X_tables = {}
+        indexer = self.get_indexer()
+        for X_name, X_table in X_tables.items():
+            new_X_tables[X_name] = pa.Table.from_arrays(
+                (
+                    indexer.obs_index(X_table["soma_dim_0"]),
+                    indexer.var_index(X_table["soma_dim_1"]),
+                    X_table[
+                        "soma_data"
+                    ].to_numpy(),  # as a side effect, consolidates chunks
+                ),
+                names=("_dim_0", "_dim_1", "soma_data"),
+            )
+        return new_X_tables
+
+
+class AxisIndexer:
```

Comment:
This can be an `attrs` class (or a `dataclass`) with the `_obs/_var_index` defaults being `None`.